domain/secretbackend/service/service.go

```diff
@@ -700,9 +700,53 @@ func (s *Service) RotateBackendToken(ctx context.Context, backendID string) erro
 	return errors.Trace(err)
 }
 
-// SetModelSecretBackend sets the secret backend for the given model UUID.
+// GetModelSecretBackend returns the secret backend name for the given model UUID, returning an error
+// satisfying [modelerrors.NotFound] if the model provided does not exist.
+func (s *Service) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, error) {
+	backendName, modelType, err := s.st.GetModelSecretBackend(ctx, modelUUID)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	switch modelType {
+	case coremodel.IAAS:
+		if backendName == provider.Internal {
+			backendName = provider.Auto
+		}
+	case coremodel.CAAS:
+		if backendName == kubernetes.BackendName {
+			backendName = provider.Auto
+		}
+	default:
+		// Should never happen.
+		return "", errors.NotValidf("model type %q", modelType)
+	}
+	return backendName, nil
+}
+
+// SetModelSecretBackend sets the secret backend config for the given model UUID,
+// returning an error satisfying [secretbackenderrors.NotFound] if the backend provided does not exist,
+// returning an error satisfying [modelerrors.NotFound] if the model provided does not exist,
+// returning an error satisfying [secretbackenderrors.NotValid] if the backend name provided is not valid.
 func (s *Service) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
-	return s.st.SetModelSecretBackend(ctx, modelUUID, backendName)
+	if backendName == "" {
+		return fmt.Errorf("missing backend name%w", errors.Hide(secretbackenderrors.NotValid))
+	}
+	if backendName == provider.Internal || backendName == kubernetes.BackendName {
+		return fmt.Errorf("secret backend name %q not valid%w", backendName, errors.Hide(secretbackenderrors.NotValid))
+	}
+	return s.st.SetModelSecretBackend(ctx, modelUUID, func(modelType coremodel.ModelType) (string, error) {
+		if backendName != provider.Auto {
+			return backendName, nil
+		}
+		switch modelType {
+		case coremodel.IAAS:
+			return provider.Internal, nil
+		case coremodel.CAAS:
+			return kubernetes.BackendName, nil
+		default:
+			return "", errors.NotValidf("model type %q", modelType)
```

Comment:
There should be a specific error for this in the domain where we don't support auto for the model type x.

Reply:
Or we should default to internal....

Reply:
No, we should not.
This really should never happen but just in case.
We should just error but not default to something.